File: samlang-checker/src/main/kotlin/samlang/checker/TypeResolution.kt
package samlang.checker

import samlang.ast.common.Type
import samlang.ast.common.Type.UndecidedType
import samlang.util.UnionFind

internal class TypeResolution : ReadOnlyTypeResolution {
    /** The union find used to manage the potential complex aliasing relation between different undecided types. */
    private val indexAliasingUnionFind: UnionFind = UnionFind()
    /**
     * A collection of known mappings between the undecided type index and the resolved types.
     * In particular:
     * - the key of the map is almost the root index in the above union find.
     * - the value of the map always represents the best knowledge of the type. i.e. we try to resolve as many undecided
     *   type as possible.
     */
    private val knownResolutions: MutableMap<Int, Type> = mutableMapOf()

    override fun toString(): String =
        "[indexAliasingUnionFind: $indexAliasingUnionFind, knownResolutions: $knownResolutions"

    /** Find the root of an index. */
    private fun Int.findRoot(): Int = indexAliasingUnionFind.find(index = this)

    /**
     * Refresh all the known mappings to ensure it represents the best knowledge we have.
     * Need to be called after each update in the knownResolutions or aliasing relation.
     */
    private fun refreshKnownMappings() {
        val keyCorrectMappings = knownResolutions.mapKeys { (key, _) -> indexAliasingUnionFind.find(index = key) }
        knownResolutions.clear()
        knownResolutions.putAll(from = keyCorrectMappings)
        knownResolutions.replaceAll { _, currentValue -> resolveType(unresolvedType = currentValue) }
    }

    override fun getPartiallyResolvedType(undecidedType: UndecidedType): Type {
        val rootIndex = undecidedType.index.findRoot()
        return knownResolutions[rootIndex] ?: UndecidedType(index = rootIndex)
    }

    override fun resolveType(unresolvedType: Type): Type =
        unresolvedType.resolveType(function = this::getPartiallyResolvedType)

    /**
     * Establish an aliasing relation between [undecidedType1] and [undecidedType2].
     *
     * It will either return the known type that both share or an error indicating there is an inconsistency.
     */
    internal fun establishAliasing(
        undecidedType1: UndecidedType,
        undecidedType2: UndecidedType,
        meet: (t1: Type, t2: Type) -> Type
    ): Type {
        val t1 = getPartiallyResolvedType(undecidedType = undecidedType1)
        val t2 = getPartiallyResolvedType(undecidedType = undecidedType2)
        if (t1 !is UndecidedType && t2 !is UndecidedType) {
            return meet(t1, t2)
        }
        val commonRoot = indexAliasingUnionFind.link(i = undecidedType1.index, j = undecidedType2.index)
        refreshKnownMappings()
        return knownResolutions[commonRoot] ?: undecidedType1
    }

    /**
     * Try to add the resolution decision for a (potentially) currently undecided type.
     *
     * It will either return an error indicating there is an inconsistency of knowledge or the best knowledge of the
     * known type.
     */
    internal fun addTypeResolution(undecidedTypeIndex: Int, decidedType: Type): Type {
        if (decidedType is UndecidedType) {
            error(message = "Use establishAliasing() instead!")
        }
        val rootOfUndecidedTypeIndex = undecidedTypeIndex.findRoot()
        val resolvedDecidedType = resolveType(unresolvedType = decidedType)
        val existingMapping = knownResolutions[rootOfUndecidedTypeIndex]
        if (existingMapping == null) {
            // If the mapping is entirely new, we can confidently use it.
            knownResolutions[rootOfUndecidedTypeIndex] = resolvedDecidedType
            // We need to refresh because we have more information.
            refreshKnownMappings()
            // Return the best knowledge we have.
            return knownResolutions[rootOfUndecidedTypeIndex]!!
        }
        return existingMapping
    }
}

File: samlang-core/checker/type-resolution.ts
import type { SamlangType, SamlangUndecidedType } from '../ast/samlang-nodes';
import { assert, checkNotNull, UnionFind } from '../utils';
import typeResolver from './type-resolver';

/** A provider of type resolution to previously undecided types. */
export interface ReadOnlyTypeResolution {
  /**
   * Given an undecided type, try to find the partially resolved type (which may still contain nested undecided
   * types) and return it. If it's not found in the resolution, return the original undecided type.
   */
  getPartiallyResolvedType(undecidedType: SamlangUndecidedType): SamlangType;

  /** Fully resolve an potentially [unresolvedType]. */
  resolveType(unresolvedType: SamlangType): SamlangType;
}

export default class TypeResolution implements ReadOnlyTypeResolution {
  /** The union find used to manage the potential complex aliasing relation between different undecided types. */
  private readonly indexAliasingUnionFind = new UnionFind();

  /**
   * A collection of known mappings between the undecided type index and the resolved types.
   * In particular:
   * - the key of the map is almost the root index in the above union find.
   * - the value of the map always represents the best knowledge of the type. i.e. we try to resolve as many undecided
   *   type as possible.
   */
  private readonly knownResolutions: Map<number, SamlangType> = new Map();

  /**
   * Refresh all the known mappings to ensure it represents the best knowledge we have.
   * Need to be called after each update in the knownResolutions or aliasing relation.
   */
  private refreshKnownMappings() {
    const keyCorrectMappings = Array.from(this.knownResolutions.entries()).map(
      ([key, currentValue]) =>
        [this.indexAliasingUnionFind.findRoot(key), this.resolveType(currentValue)] as const
    );
    this.knownResolutions.clear();
    keyCorrectMappings.forEach(([key, value]) => {
      this.knownResolutions.set(key, value);
    });
  }

  getPartiallyResolvedType = (undecidedType: SamlangUndecidedType): SamlangType => {
    const rootIndex = this.indexAliasingUnionFind.findRoot(undecidedType.index);
    return this.knownResolutions.get(rootIndex) ?? { type: 'UndecidedType', index: rootIndex };
  };

  resolveType = (unresolvedType: SamlangType): SamlangType =>
    typeResolver(unresolvedType, this.getPartiallyResolvedType);

  /**
   * Establish an aliasing relation between [undecidedType1] and [undecidedType2].
   *
   * It will either return the known type that both share or an error indicating there is an inconsistency.
   */
  establishAliasing(
    undecidedType1: SamlangUndecidedType,
    undecidedType2: SamlangUndecidedType,
    meet: (t1: SamlangType, t2: SamlangType) => SamlangType
  ): SamlangType {
    const t1 = this.getPartiallyResolvedType(undecidedType1);
    const t2 = this.getPartiallyResolvedType(undecidedType2);
    if (t1.type !== 'UndecidedType' && t2.type !== 'UndecidedType') {
      return meet(t1, t2);
    }
    const commonRoot = this.indexAliasingUnionFind.link(undecidedType1.index, undecidedType2.index);
    this.refreshKnownMappings();
    return this.knownResolutions.get(commonRoot) ?? undecidedType1;
  }

  /**
   * Try to add the resolution decision for a (potentially) currently undecided type.
   *
   * It will either return an error indicating there is an inconsistency of knowledge or
   * the best knowledge of the known type.
   */
  addTypeResolution(undecidedTypeIndex: number, decidedType: SamlangType): SamlangType {
    assert(decidedType.type !== 'UndecidedType', 'Use establishAliasing() instead!');

    const rootOfUndecidedTypeIndex = this.indexAliasingUnionFind.findRoot(undecidedTypeIndex);
    const resolvedDecidedType = this.resolveType(decidedType);
    const existingMapping = this.knownResolutions.get(rootOfUndecidedTypeIndex);
    if (existingMapping == null) {
      // If the mapping is entirely new, we can confidently use it.
      this.knownResolutions.set(rootOfUndecidedTypeIndex, resolvedDecidedType);
      // We need to refresh because we have more information.
      this.refreshKnownMappings();
      // Return the best knowledge we have.
      const resolved = checkNotNull(this.knownResolutions.get(rootOfUndecidedTypeIndex));
      return resolved;
    }
    return existingMapping;
  }
}
